File: java/src/edu/caltech/ipac/firefly/visualize/ui/color/Color.java
/*
	Copyright 2008 Marco Mustapic
	
    This file is part of Agilar GWT Widgets.

    Agilar GWT Widgets is free software: you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    Agilar GWT Widgets is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with Agilar GWT Widgets.  If not, see <http://www.gnu.org/licenses/>.
*/

package edu.caltech.ipac.firefly.visualize.ui.color;

public class Color {
    private static final double FACTOR = 0.7;

	public static void HSVToRGB(float [] hsv, float [] rgb)
	{
		
		float r = 0.0f;
		float g = 0.0f;
		float b = 0.0f;
		float h = hsv[0];
		float s = hsv[1];
		float v = hsv[2];
		
		float hf = h / 60.0f;
		int i = (int) Math.floor(hf);
		float f = hf - i;
		float pv = v * (1 - s);
		float qv = v * (1 - s * f);
		float tv = v * (1 - s * (1 - f));
		
		switch (i)
		{
		// Red is the dominant color
		case 0:
			r = v;
			g = tv;
			b = pv;
			break;
			
			// Green is the dominant color
		case 1:
			r = qv;
			g = v;
			b = pv;
			break;
		case 2:
			r = pv;
			g = v;
			b = tv;
			break;
			
			// Blue is the dominant color
		case 3:
			r = pv;
			g = qv;
			b = v;
			break;
		case 4:
			r = tv;
			g = pv;
			b = v;
			break;
			
			// Red is the dominant color
		case 5:
			r = v;
			g = pv;
			b = qv;
			break;
			
			// Just in case we overshoot on our math by a little, we put these here. Since its a switch it won't slow us down at all to put these here.
		case 6:
			r = v;
			g = tv;
			b = pv;
			break;
		case -1:
			r = v;
			g = pv;
			b = qv;
			break;
		}
		
		rgb[0] = r;
		rgb[1] = g;
		rgb[2] = b;
	}

	public static void RGBToHSV(float [] rgb, float[] hsv)
	{
		float r = rgb[0];
		float g = rgb[1];
		float b = rgb[2];
		float h, s, v;
		
		float min = min3(r, g, b);
		float max = max3(r, g, b);
		
		// calculate value
		v = max; // value
		if (v == 0)
		{
			h = 0;
			s = 0;
		}
		
		// calculate saturation
		s = max - min;
		if (s == 0)
		{
			h = 0;
		}
		
		// calculate hue
		if (max == r)
		{
			h = 0.0f + 60.0f * (g - b);
			if (h < 0.0)
			{
				h += 360.0f;
			}
		}
		else if (max == g)
		{
			h = 120.0f + 60.0f * (b - r);
		}
		else
		/* max == b */{
			h = 240.0f + 60.0f * (r - g);
		}
		
		hsv[0] = h;
		hsv[1] = s;
		hsv[2] = v;
	}

	public static float min3(float v1, float v2, float v3)
	{
		float min;

		min = v1 < v2 ? v1 : v2;
		min = min < v3 ? min : v3;

		return min;
	}

	public static float max3(float v1, float v2, float v3)
	{
		float max;

		max = v1 > v2 ? v1 : v2;
		max = max > v3 ? max : v3;

		return max;
	}
	
	public static String toHex(int [] rgb)
	{
		String hex = "";
		hex += rgb[0]>=16?Integer.toHexString(rgb[0]):"0"+Integer.toHexString(rgb[0]);
		hex += rgb[1]>=16?Integer.toHexString(rgb[1]):"0"+Integer.toHexString(rgb[1]);
		hex += rgb[2]>=16?Integer.toHexString(rgb[2]):"0"+Integer.toHexString(rgb[2]);
		
		return hex;
	}
	
	public static void toInt(float [] frgb, int [] irgb)
	{
		irgb[0] = Math.round(frgb[0] * 255.0f);
		irgb[1] = Math.round(frgb[1] * 255.0f);
		irgb[2] = Math.round(frgb[2] * 255.0f);
	}

	public static void toFloat(int [] irgb, float [] frgb)
	{
		frgb[0] = ((float)irgb[0]) / 255.0f;
		frgb[1] = ((float)irgb[1]) / 255.0f;
		frgb[2] = ((float)irgb[2]) / 255.0f;
	}

    public static boolean isHexColor(String text) {
        boolean retval= false;
        if (text.length() == 6 || (text.length()==7 && text.startsWith("#"))) {
            if (text.startsWith("#")) text=text.substring(1);
            retval= true;
            for (int i = 0; i < text.length(); i++) {
                char c = text.charAt(i);
                if (Character.digit(c, 16) == -1) {
                    retval= false;
                    break;
                }
            }
        }
        return retval;
    }

    public static int [] toRGB(String colorStr) {
        int retval[]= null;
        if (isHexColor(colorStr)) {
            try {
                //Note: subString(1) returned a new instance, so isHexColor(colorStr) cannot remove "#" in colorStr.
                if (colorStr.startsWith("#")) colorStr=colorStr.substring(1);
                String rStr= colorStr.substring(0,2);
                String gStr= colorStr.substring(2,4);
                String bStr= colorStr.substring(4);
                int red= Integer.parseInt(rStr, 16);
                int green= Integer.parseInt(gStr, 16);
                int blue= Integer.parseInt(bStr, 16);
                retval= new int[] { red, green, blue};
            } catch (NumberFormatException e) {
                retval= null;
            }
        }
        return retval;
    }

    public static String brighter(String colorStr) {
        return brighter(colorStr,FACTOR);

    }


    public static String brighter(String colorStr,double factor) {
        int rgb[]= toRGB(colorStr);
        int r = rgb[0];
        int g = rgb[1];
        int b = rgb[2];

        /* From 2D group:
         * 1. black.brighter() should return grey
         * 2. applying brighter to blue will always return blue, brighter
         * 3. non pure color (non zero rgb) will eventually return white
         */
        int i = (int)(1.0/(1.0-factor));
        if ( r == 0 && g == 0 && b == 0) {
            return toHex(new int[] {Math.min((int)(i/factor), 255),
                                    Math.min((int)(i/factor), 255),
                                    Math.min((int)(i/factor), 255) } );
        }
        if ( r > 0 && r < i ) r = i;
        if ( g > 0 && g < i ) g = i;
        if ( b > 0 && b < i ) b = i;

        return toHex(new int[] {Math.min((int)(r/factor), 255),
                                Math.min((int)(g/factor), 255),
                                Math.min((int)(b/factor), 255) } );
    }

    /**
     * Creates a new <code>Color</code> that is a darker version of this
     * <code>Color</code>.
     * <p>
     * This method applies an arbitrary scale factor to each of the three RGB
     * components of this <code>Color</code> to create a darker version of
     * this <code>Color</code>.  Although <code>brighter</code> and
     * <code>darker</code> are inverse operations, the results of a series
     * of invocations of these two methods might be inconsistent because
     * of rounding errors.
     * @return  a new <code>Color</code> object that is
     *                    a darker version of this <code>Color</code>.
     * @see        java.awt.Color#brighter
     * @since      JDK1.0
     */
    public static String darker(String colorStr) {
        return darker(colorStr,FACTOR);
    }

    public static String darker(String colorStr, double factor) {
        int rgb[]= toRGB(colorStr);
        int r = rgb[0];
        int g = rgb[1];
        int b = rgb[2];
        return toHex(new int[] {Math.max((int)(r *factor), 0),
                                Math.max((int)(g *factor), 0),
                                Math.max((int)(b *factor), 0)});
    }


    /**
     * Make a simple color map array of no more than 10 entries.  Map goes from darker to brighter.
     * @param baseColor
     * @param mapSize
     * @return
     */
    public static String[] makeSimpleColorMap(String baseColor, int mapSize) {
        String c[]= null;
        if (mapSize<=1) {
            return new String[] {baseColor};
        }
        if (Color.isHexColor(baseColor)) {
            if (mapSize>10) mapSize= 10;

            int baseIdx= mapSize/5;

            int rgb[]= Color.toRGB(baseColor);
            int maxCnt= 0;
            int minCnt= 0;
            for(int idx : rgb) {
                if (idx>250) maxCnt++;
                if (idx<5)   minCnt++;
            }
            if ((maxCnt==1 && minCnt==2) ||(maxCnt==2 && minCnt==1) || maxCnt==3) {
                baseIdx= mapSize-1;
            }
            else if (minCnt==3) {
                baseIdx= 0;
            }


            c= new String[mapSize];
            double factor= (mapSize>5) ? .9 : .85;

            c[baseIdx]= baseColor;
            for(int i= baseIdx+1; (i<mapSize); i++) {
                c[i]= Color.brighter(c[i-1],factor);
            }
            for(int i= baseIdx-1; (i>=0); i--) {
                c[i]= Color.darker(c[i+1], factor);
            }
        }
        return c;
    }

}

File: src/firefly/js/util/Color.js
import chroma from 'chroma-js';
import {isArray} from 'lodash';

export default { makeSimpleColorMap, getBWBackground, shadeColor};

export const toRGBA= (color, alpha=1) => chroma.valid(color) ? chroma(color).alpha(alpha).rgba() : [0,0,0,alpha];
export const toRGB= (color) => chroma.valid(color) ? chroma(color).rgb() : [0,0,0];

/**
 * @summary extract R,G,B amd alpha value from color string like 'rgba(...)', 'rgb(...)' or '#rrggbb'
 * @param color
 * @returns {Array} [r, g, b, a]
 */
export const getRGBA= (color)  => chroma.valid(color) ? chroma(color).rgba() : [0,0,0,1];


export const brighter = (colorStr,level=1) =>
    chroma.valid(colorStr) ? chroma(colorStr).brighten(level).toString() : undefined;

/**
 * Creates a new <code>Color</code> that is a darker version of this
 * <code>Color</code>.
 * <p>
 * This method applies an arbitrary scale factor to each of the three RGB
 * components of this <code>Color</code> to create a darker version of
 * this <code>Color</code>.  Although <code>brighter</code> and
 * <code>darker</code> are inverse operations, the results of a series
 * of invocations of these two methods might be inconsistent because
 * of rounding errors.
 * @param colorStr
 * @param [level]
 * @return {*} a new <code>Color</code> object that is
 *                    a darker version of this <code>Color</code>.
 * @since      JDK1.0
 */
export const darker = (colorStr,level=1) =>
    chroma.valid(colorStr) ? chroma(colorStr).darken(level).toString() : undefined;

/**
 * Make a simple color map array of no more than 10 entries.  Map goes from darker to brighter.
 * @param baseColor
 * @param mapSize
 * @param reverse
 * @return {*}
 */
function makeSimpleColorMap(baseColor, mapSize=10, reverse=false) {
    const c1= chroma(baseColor).brighten(2);
    const c2= chroma(baseColor).darken(1);
    const cAry= reverse ? [c2,c1] : [c1,c2];
   return chroma.scale(cAry).colors(mapSize);
}

const [R, G, B, A] = [0, 1, 2, 3];
const [LUMI, CONTRAST, HSP] = [0, 1, 2];



export function rateOpacity(color, ratio) {
    const rgba = getRGBA(color);
    if (!rgba) {
        return 'rgba(255, 255, 255,' + (1.0*ratio) + ')';
    }
    const newAlpha = Math.max(Math.min(rgba[3] * ratio, 1.0), 0.0);
    const newColor = rgba.slice(R, A);
    newColor.push(newAlpha);

    return toRGBAString(newColor);
}

export const maximizeOpacity= (color) => chroma.valid(color) ? chroma(color).alpha(1).toString() : 'white';

/**
 * @summary get contrast color in terms of black or white, based on various algorithms, relative luminance, contrast, or HSP color model.
 * @param {string} color
 * @param {Number} method
 * @returns {*}
 */
export function getBWBackground(color, method = LUMI) {
    let weight;
    let th;
    let luma;

    const rgba = getRGBA(color);
    if (!rgba) {
        return 'rgba(255, 255, 255, 1.0)';
    }

    const linearLuma = (w) => {
        return [R, G, B].reduce((lum, c) => {
            lum += w[c] * rgba[c];
            return lum;
        }, 0.0);
    };
    const squareLuma = (w) => {
        return [R, G, B].reduce((lum, c) => {
            lum += w[c] * rgba[c] * rgba[c];
            return lum;
        }, 0.0);
    };


    switch (method) {
        case CONTRAST:
        case HSP:
            weight = [0.299, 0.587, 0.114];
            th = 128;

            if (method === CONTRAST) {
                luma =  linearLuma(weight);
            } else {
                luma = Math.sqrt(squareLuma(weight));

            }
            break;
        case LUMI:
        default:
            weight = [0.2126, 0.7152, 0.0722];
            th = 165;

            luma = linearLuma(weight);
            break;
    }
    const bkGrey = luma < th ? parseInt('ff', 16) : 0;  // dark color -> get bright, bright color -> get dark

    return `rgba(${bkGrey}, ${bkGrey}, ${bkGrey}, 1.0)`;
}

/**
 * @summary get complementary color
 * @param {string} color
 * @returns {*}
 */
function getComplementaryColor(color) {
    const rgba = getRGBA(color);

    if (!rgba)  return 'rgba(201, 201, 201, 1.0)';   //'#e3e3e3'
    let rgb = rgba.slice(0, 3);

    const RGB2HSV = (rgb) =>
    {
        const hsv = {};
        const maxV = Math.max(...rgb);
        const minV = Math.min(...rgb);

        const dif = maxV - minV;

        hsv.saturation = (maxV === 0.0) ? 0 : (100.0 * dif / maxV);
        if (hsv.saturation === 0) {
            hsv.hue = 0;
        } else if (rgb[R] === maxV) {
            hsv.hue = 60.0 * (rgb[G] - rgb[B]) / dif;
        } else if (rgb[G] === maxV) {
            hsv.hue = 120.0 + 60.0 * (rgb[B] - rgb[R]) / dif;
        } else if (rgb[B] === maxV) {
            hsv.hue = 240.0 + 60.0 * (rgb[R] - rgb[G]) / dif;
        }
        if (hsv.hue < 0.0) hsv.hue += 360.0;
        hsv.value = Math.round(maxV * 100 / 255);
        hsv.hue = Math.round(hsv.hue);
        hsv.saturation = Math.round(hsv.saturation);
        return hsv;
    };

    const HSV2RGB = (hsv) => {
        let rgb = [0, 0, 0];
        const toRGB = (rgb, r, g, b) => {
            rgb[R] = r;
            rgb[G] = g;
            rgb[B] = b;
            return rgb.map((v) => (Math.round(v*255)));
        };

        if (hsv.saturation === 0) {
            let v = Math.round(hsv.value*2.55);

            if (hsv.hue === 180.0) {
                v = 255 - v;
            }
            rgb = toRGB(rgb, v, v, v);
        } else {
            hsv.hue /= 60.0;
            hsv.saturation /= 100.0;
            hsv.value /= 100.0;


            const i = Math.floor(hsv.hue);
            const f = hsv.hue - i;
            const a = hsv.value * (1.0 - hsv.saturation);
            const b = hsv.value * (1.0 - hsv.saturation * f);
            const c = hsv.value * (1.0 - hsv.saturation * (1-f));
            const d = hsv.value;
            switch(i) {
                case 0:
                    rgb = toRGB(rgb, d, c, a);
                    break;
                case 1:
                    rgb = toRGB(rgb, b, d, a);
                    break;
                case 2:
                    rgb = toRGB(rgb, a, d, c);
                    break;
                case 3:
                    rgb = toRGB(rgb, a, b, d);
                    break;
                case 4:
                    rgb = toRGB(rgb, c, a, d);
                    break;
                case 5:
                    rgb = toRGB(rgb, d, a, b);
                    break;
                default:
                    rgb = toRGB(rgb, 0, 0, 0);
                    break;
            }
        }
        return rgb;
    };

    const HueShift  = (h, s)  => {
        h += s;

        while (h >= 360.0) {
            h -= 360.0;
        }
        while (h < 0.0) {
            h+= 360.0;
        }
        return h;
    };

    const tmphsv = RGB2HSV(rgb);
    tmphsv.hue = HueShift(tmphsv.hue, 180.0);
    rgb = HSV2RGB(tmphsv);

    return `rgba(${rgb[R]}, ${rgb[G]}, ${rgb[B]}, ${rgba[3]})`;
}

export function toRGBAString(rgba) {
    const len = (!rgba || !isArray(rgba)) ? 0 : rgba.length;
    const val = len === 0 ? 255 : rgba[len-1];

    if (len <= A) {
        if (len === 0) rgba = [];
        for (let i = 0; i < A - len; i++ ) {
            rgba.push(val);
        }
        rgba.push(1.0);
    }

    return `rgba(${rgba[R]}, ${rgba[G]}, ${rgba[B]}, ${rgba[A]})`;
}

/*
 * @param {String} color - hex color, exactly seven characters log, starting with '#'
 * @param {Number} percentage (0.1 means 10 percent lighter, -0.1 - 10 percent darker)
 * @return {String} lighter or darker shade of the given hex color
 * from http://stackoverflow.com/questions/5560248/programmatically-lighten-or-darken-a-hex-color-or-rgb-and-blend-colors
 */
function shadeColor(color, percent) {
    const [R, G, B] = toRGB(color.slice(1));
    const t=percent<0?0:255,p=percent<0?percent*-1:percent;
    return `#${(0x1000000+(Math.round((t-R)*p)+R)*0x10000+(Math.round((t-G)*p)+G)*0x100+(Math.round((t-B)*p)+B)).toString(16).slice(1)}`;
}

